# Part 1. Equation of a Slime

How many late days are you using for this assignment? 0


In [126]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
# Imports section


## 1. Loading the dataset

In [127]:
# Using pandas load the dataset
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)

data = pd.read_csv('science_data_large.csv')
display(data.head(15))
display(data.info())

,Temperature °C,Mols KCL,Size nm^3
0,469,647,6.244743e+05
1,403,694,5.779610e+05
2,302,975,6.196847e+05
3,779,916,1.460449e+06
4,901,18,4.325726e+04
5,545,637,7.124634e+05
6,660,519,7.006960e+05
7,143,869,2.718260e+05
8,89,461,8.919803e+04
9,294,776,4.770210e+05


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 non-null   float64
dtypes: float64(1), int64(2)
memory usage: 23.6 KB


None

## 2. Splitting the dataset

In [128]:
# Take the pandas dataset and split it into our features (X) and label (y)

X = data.drop('Size nm^3', axis=1)
y = data['Size nm^3']

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import StandardScaler


X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.1, random_state = 42)

## 3. Perform a Linear Regression

In [129]:
# Use sklearn to train a model on the training set
# Create a sample datapoint and predict the output of that sample with the trained model
model = LinearRegression()
model.fit(X_train, y_train)

sample = X_test[0:1]
predict = model.predict(sample)
print(f"Predicted ouput: {predict}")
# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means
score = model.score(X_test, y_test)
print(f"Model score: {score:.5f}")
# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX

coeffs = model.coef_
intercept = model.intercept_

print("Coefficients:")
for name, coeff in zip(X.columns, coeffs):
	print(f"Coefficient for {name}: {coeff:.5f}")
print(f"Intercept: {intercept:.5f}")

Predicted ouput: [235911.1927226]
Model score: 0.85525
Coefficients:
Coefficient for Temperature °C: 866.14641
Coefficient for Mols KCL: 1032.69507
Intercept: -409391.47958


Write the linear equation of a slime: (example equation: $E = mc^2$)

$Y = 866.14641 \cdot \text{(Temperature)} + 1032.69507 \cdot \text{(Mols KCL)} - 409391.47958$


Report on score and explain meaning:
The model score was .85, which shows how well the model is when it comes to the testing. The closer it is to 1 the better, if the model score is = 1 it means that it is a perfect indication.

## 4. Use Cross Validation

In [130]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42
from sklearn.model_selection import cross_val_score, KFold

# Define the cross-validation strategy
kf = KFold(n_splits=5, shuffle=True, random_state=42)

# Perform cross-validation
scores = cross_val_score(model, X, y, cv=kf, scoring='r2')

# Report on their finding and their significance
print(f"Cross validation scores: {scores}")
print(f"Mean score: {scores.mean():.5f}")


Cross validation scores: [0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]
Mean score: 0.85973


Write findings here:

The folding for the cross validation scores helps with in assesing how the model generalizes new data to prevent and identify overfitting or underfitting. The shuffling ensures that the data is shuffled in between folds and the random state adds extra variables so that it can be reproduced. The scoring is specifies R^2 which is the scoring system to evaluate the accuracy of the model. 

## 5. Using Polynomial Regression

In [131]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)

from sklearn.preprocessing import PolynomialFeatures
from sklearn.model_selection import KFold
from sklearn.pipeline import Pipeline

pipeline = Pipeline([
    ('poly', PolynomialFeatures(degree=2)),
    ('model', LinearRegression())
])

kf = KFold(n_splits=5, shuffle=True, random_state=42)
scores = cross_val_score(pipeline, X, y, cv=kf, scoring='r2')

print(f"Cross validation scores with polynomial features: {scores}")
print(f"Mean score: {scores.mean()}")

# Fit the pipeline on the entire dataset to get the coefficients
pipeline.fit(X, y)

model = pipeline.named_steps['model']

coeffs = model.coef_
intercept = model.intercept_

# Extract feature names for polynomial features
poly_feature_names = pipeline.named_steps['poly'].get_feature_names_out(X.columns)

print("Coefficients:")
for name, coeff in zip(poly_feature_names, coeffs):
    print(f"Coefficient for {name}: {coeff:.5f}")
print(f"Intercept: {intercept:.5f}")


Cross validation scores with polynomial features: [1. 1. 1. 1. 1.]
Mean score: 1.0
Coefficients:
Coefficient for 1: 0.00000
Coefficient for Temperature °C: 12.00000
Coefficient for Mols KCL: -0.00000
Coefficient for Temperature °C^2: -0.00000
Coefficient for Temperature °C Mols KCL: 2.00000
Coefficient for Mols KCL^2: 0.02857
Intercept: 0.00002


Write the polynomial equation of a slime: (example equation: $E = mc^2$)

$y = 0.00002 + 12.00000 \cdot \text{Temperature °C} - 0.00000 \cdot \text{Mols KCL} - 0.00000 \cdot (\text{Temperature °C})^2 + 2.00000 \cdot \text{Temperature °C} \cdot \text{Mols KCL} + 0.02857 \cdot (\text{Mols KCL})^2$


Report on the score and interpret:
The score is a 1 for every cross validation score, and the mean is also 1, meaning that the Polynomial Regresssion is the perfect model for this data set.

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [132]:
# Load the dataset. Then train and evaluate the classification models.

data = pd.read_csv('ckd_feature_subset.csv')
display(data.head(15))
display(data.info())

from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.preprocessing import StandardScaler

models = {"Logistic Regression": LogisticRegression(),
    "Support Vector Classifier": SVC(),
    "k-Nearest Neighbors": KNeighborsClassifier(),
}

results = []
X = data.drop('Target_ckd', axis=1)
y = data['Target_ckd']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)


# Perform 5-fold cross-validation for each model
for model_name, model in models.items():
    cv_scores = cross_val_score(model, X, y, cv=5)
    results.append({
        "Model": model_name,
        "Average Accuracy": np.mean(cv_scores),
        "Standard Deviation": np.std(cv_scores)
    })

# Convert results to a DataFrame
results_df = pd.DataFrame(results)
print(results_df)




,age,bp,wbcc,appet_poor,appet_good,rbcc,Target_ckd
0,0.688312,0.333333,0.000000,1,0,0.000000,1
1,0.545455,0.333333,0.128319,1,0,0.305085,1
2,0.714286,0.500000,0.238938,1,0,0.186441,1
3,0.688312,0.333333,0.283186,0,1,0.338983,1
4,0.441558,0.333333,0.221239,1,0,0.220339,1
5,0.740260,0.833333,0.265487,0,1,0.355932,1
6,0.870130,0.333333,0.668142,0,1,0.237288,1
7,0.870130,0.500000,0.150442,0,1,0.372881,1
8,0.194805,0.666667,0.380531,0,1,0.305085,1
9,0.753247,0.833333,0.163717,0,1,0.220339,1


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 153 entries, 0 to 152
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   age         153 non-null    float64
 1   bp          153 non-null    float64
 2   wbcc        153 non-null    float64
 3   appet_poor  153 non-null    int64  
 4   appet_good  153 non-null    int64  
 5   rbcc        153 non-null    float64
 6   Target_ckd  153 non-null    int64  
dtypes: float64(4), int64(3)
memory usage: 8.5 KB


None

                       Model  Average Accuracy  Standard Deviation
0        Logistic Regression          0.856129            0.054236
1  Support Vector Classifier          0.928172            0.059993
2        k-Nearest Neighbors          0.928172            0.031893


## Results and Conclusion for Classification Experiments

The results of the three models, are that out of the three, the logistic regression model had the worst average acuracy with a .05 std. The SVC and the k-nearest had the same exact acuraccy but the k-nearest was better because it had a smaller standard deviation. 

In [133]:
# Experiments with Neural Network.

nn_configs = {
    "NN-Config 1": MLPClassifier(hidden_layer_sizes=(10,),activation='relu', max_iter=500, random_state=42),
    "NN-Config 2": MLPClassifier(hidden_layer_sizes=(10,),activation='logistic', max_iter=500, random_state=42),
    "NN-Config 3": MLPClassifier(hidden_layer_sizes=(10, 10), activation='relu', max_iter=500, random_state=42),
    "NN-Config 4": MLPClassifier(hidden_layer_sizes=(10, 10), activation='logistic', max_iter=500, random_state=42),}

results = []
# Perform 5-fold cross-validation for each Neural Network configuration
for config_name, nn_model in nn_configs.items():
    cv_scores = cross_val_score(nn_model, X, y, cv=5)
    results.append({
        "Model": config_name,
        "Average Accuracy": np.mean(cv_scores),
        "Standard Deviation": np.std(cv_scores)
    })

# Convert results to DataFrame
results_df = pd.DataFrame(results)

# Print the final results including NN configurations
print(results_df)

/Users/<redacted>/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/Users/<redacted>/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/Users/<redacted>/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/Users/<redacted>/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/Users/<redacted

         Model  Average Accuracy  Standard Deviation
0  NN-Config 1          0.849677            0.054005
1  NN-Config 2          0.771398            0.034312
2  NN-Config 3          0.927957            0.038746
3  NN-Config 4          0.725591            0.013368


/Users/<redacted>/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(


## Results and Conclusion for Neural Network Experiments

I ran 4 different models, two with relu activation and two with logistic activation. The relu activations had better accuracy than the logistic activations. I also changed the layer sizes. NN1 and NN3 only had 1 single hidden layer with 10 neurons, and NN2 and NN4 had 2 single layers with 10 neurons each. With each layer, the model becomes more accurate as shown above. As well as relu activation is more accurate than logistic activation for this data set. The relu had more accuracy with more layers, but the logistic one had less acuraccy with more layers but a smaller standard deviation. Overall the relu activation had the best accuracy. 